### PR TITLE
Update deprecated Vue slot syntax

### PR DESCRIPTION
**Pull request by wxwxwxwx9**
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [ ] Feature addition or enhancement
- [x] Code maintenance
- [ ] Others, please explain:

This PR addresses part of #1078 in removing the deprecated Vue slot syntax.

**Overview of changes:**
Update all the internal Vue slot syntax to v-slot (e.g. `slot="example"` to `v-slot:example`)

**Anything you'd like to highlight / discuss:**
Since the changes are coupled with the processing of the nodes, we have to be careful in ensuring that existing features are not affected. While I have ensured that all the existing test cases are passing, I'm not sure if this change will break some other existing features. Would be great if I can get some advice here :-/

There is, however, one area that I identified to have regressed but was not able to pinpoint the actual cause of it from my changes. 

Given the following, the item within the slot currently are not appearing as the panel header. I'm still looking into this issue. 
```
<panel  type="info" alt="hi">
  <p slot="header" class="card-title">
    <i><strong>
      <span style="color:#FF0000;">R</span>
      <span style="color:#FF7F00;">A</span>
      <span style="color:#FFFF00;">I</span>
      <span style="color:#00FF00;">N</span>
      <span style="color:#0000FF;">B</span>
      <span style="color:#4B0082;">O</span>
      <span style="color:#9400D3;">W</span>
    </strong></i>
  </p>
</panel>
```

**Testing instructions:**
`npm run test`

**Proposed commit message: (wrap lines at 72 characters)**
Update deprecated Vue slot syntax. 

Our internal Vue slot syntax are deprecated according to official 
Vue documentation. To ensure our codebase is up-to-date, let's 
convert all the deprecated syntax to the latest Vue slot shorthand 
syntax that was introduced in version 2.6.0. 

We will still support the old slot syntax for backward compability, 
by processing and converting the old slot syntax to the new 
shorthand syntax. 

Externally, users will be able to use the latest shorthand syntax for
slots as well, as our processing targets the latest shorthand syntax. 

---

**Checklist:** :ballot_box_with_check:

- [x] Updated the documentation for feature additions and enhancements
- [x] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No blatantly unrelated changes
- [x] Pinged someone for a review!

### CODE REVIEW

File: packages/core/src/plugins/algolia.js [hunk 1]
```diff
@@ -24,9 +24,9 @@ function addNoIndexClasses(content) {
     'dropdown',
     'b-modal',
     'panel:not([expanded])',
-    'popover div[slot=content]',
-    'question div[slot=hint]',
-    'question div[slot=answer]',
+    'popover div[v-slot\\:content]',
```
Comment (wxwxwxwx9) [inline packages/core/src/plugins/algolia.js:27]: @ang-zeyu can I check if these actually should be updated?

File: packages/core/test/unit/html/NodeProcessor.test.js [hunk 2]
```diff
@@ -20,6 +20,8 @@ const processAndVerifyTemplate = (template, expectedTemplate, postProcess = fals
     const nodeProcessor = getNewDefaultNodeProcessor();
 
     if (postProcess) {
+      // need to process node first (convert slot to v-slot) before doing post-processing
+      dom.forEach(node => nodeProcessor.processNode(node, new Context(path.resolve(''))));
```
Comment (wxwxwxwx9) [inline packages/core/test/unit/html/NodeProcessor.test.js:24]: Post-processing involves adding of panel id so we have to process the node first before doing post-processing.

**Review by ang-zeyu — COMMENTED** [review]
just some initial comments:

---

>There is, however, one area that I identified to have regressed but was not able to pinpoint the actual cause of it from my changes.
>
>Given the following, the item within the slot currently are not appearing as the panel header. I'm still looking into this issue.

Likely this:

(vue docs)
> Note that v-slot can only be added to a <template> (with one exception), unlike the deprecated slot attribute.

File: packages/core/src/Page/index.js [hunk 3]
```diff
@@ -194,15 +194,15 @@ class Page {
       }
 
       // Check if heading / panel is under a panel's header slots, which is handled specially below.
-      const slotParents = $(elem).parentsUntil(context).filter('[slot="header"]').not(elem);
+      const slotParents = $(elem).parentsUntil(context).filter('[v-slot\\:header]').not(elem);
```
[the remark this answers is not in the document]
Comment (wxwxwxwx9) [reply packages/core/src/Page/index.js]: Noted! Made the changes :-)

**Comment by wxwxwxwx9** [comment]
@ang-zeyu Thanks so much for the review!

I have made the changes accordingly and tested as much as I could on my side. I think it should work fine. I've verified that the test cases failing in the CI currently are expected due to the changes I've made. Let me know if there happens to be any edge case that I missed out. 

I'm thinking of adding more test cases to our test suite for NodeProcessor to ensure that both deprecated syntax and the new shorthand syntax would work. Should I go ahead with this? 

If so, I'm not sure if I should create test cases for every component since it seems a little overkill. I'm thinking of just creating for a few components (not sure which ones I should pick though, any of them should do in my opinion). What do you think? 

> > There is, however, one area that I identified to have regressed but was not able to pinpoint the actual cause of it from my changes.
> > Given the following, the item within the slot currently are not appearing as the panel header. I'm still looking into this issue.
> 
> Likely this:
> 
> (vue docs)
> 
> > Note that v-slot can only be added to a  (with one exception), unlike the deprecated slot attribute

Thanks for highlighting this to me. Managed to resolve it :-) what I did was to convert all tag names to "template" if there is a shorthand vue slot syntax in the node.

File: packages/core/src/html/NodeProcessor.js [hunk 4]
```diff
@@ -88,17 +89,70 @@ class NodeProcessor {
...
+                Both syntax's slot names match what we are looking for.
+                We will use the the shorthand syntax and delete the deprecated syntax.
+              */
+              delete child.attribs.slot;
+            }
+          }
+          if (isTargetVslotShorthand) {
+            /*
+              User may use tag names like "div" or "p" but we have to change it to "template"
+              for vslot shorthand syntax to work, according to Vue2.6 syntax documenatation.
```
Comment (ang-zeyu) [inline packages/core/src/html/NodeProcessor.js:136]: you'll need to "shift it one level deeper" in the dom tree instead (then wrap with `<template #xx>`) as the `<p slot="..."` syntax dosen't erase the `p` element.

File: packages/core/src/html/NodeProcessor.js [hunk 5]
```diff
@@ -123,12 +177,17 @@ class NodeProcessor {
...
     node.children.forEach((child) => {
-      const slot = child.attribs && child.attribs.slot;
-      if (slot) {
-        // Turns <div slot="content">... into <div data-mb-slot-name=content>...
-        child.attribs['data-mb-slot-name'] = slot;
-        delete child.attribs.slot;
+      // Turns <div #content>... into <div data-mb-slot-name=content>...
+      const vslotShorthand = NodeProcessor.getVslotShorthand(child);
+      if (vslotShorthand) {
+        const slotName = vslotShorthand.substring(1, vslotShorthand.length); // remove # from vslot shorthand
```
Comment (ang-zeyu) [inline packages/core/src/html/NodeProcessor.js:188]: this seems repeated quite often. how about moving it into `getVslotShorthand`?

File: packages/core/src/html/NodeProcessor.js [hunk 6]
```diff
@@ -88,17 +89,70 @@ class NodeProcessor {
...
+          if (isTargetVslotShorthand) {
+            /*
+              User may use tag names like "div" or "p" but we have to change it to "template"
+              for vslot shorthand syntax to work, according to Vue2.6 syntax documenatation.
+            */
+            child.name = 'template';
+            return true;
+          }
+        }
+        if (hasSlot) {
```
Comment (ang-zeyu) [inline packages/core/src/html/NodeProcessor.js:142]: how about:

- at every element (can maintain a set) (`processNode`) that requires slot processing
   - check if its children has `slot="..."`, transform to new syntax if so (e.g. a `transformOldSlotSyntax(node)` method)
   - shift its slot children deeper one level if `node.name !== 'template'` (previous comment) (similarly something like `shiftSlotDeeper(node)`?)
  
that way:
- you'll only need to maintain one flow (your new code targetting `#slotName`)
- the old tests using `slot="xx"` also follow this flow, so we only need to include a simple test for old slot syntax transformation

**Review by ang-zeyu — COMMENTED** [review]
Mostly looks good! 🙂

Just a couple suggestions (one to remove the old slot processing code entirely) and one issue:

File: packages/core/src/html/NodeProcessor.js (hunk 6 above)
Comment (wxwxwxwx9) [reply packages/core/src/html/NodeProcessor.js]: I've made the changes but I'm not sure if it's exactly what you're suggesting :O let me know if I have misinterpreted something (especially regarding the set and not having duplicate function calls)!

File: packages/core/src/html/NodeProcessor.js (hunk 5 above)
Comment (wxwxwxwx9) [reply packages/core/src/html/NodeProcessor.js]: Hmm yup that's possible!

File: packages/core/src/html/NodeProcessor.js [hunk 7]
```diff
@@ -204,8 +223,14 @@ class NodeProcessor {
...
    */
   static _assignPanelId(node) {
-    const slotChildren = node.children && node.children.filter(child => _.has(child.attribs, 'slot'));
-    const headerSlot = slotChildren.find(child => child.attribs.slot === 'header');
+    const slotChildren = node.children
+      && node.children.filter(child => NodeProcessor.getVslotShorthandName(child) !== '');
+
+    const headerSlot = slotChildren.find((child) => {
+      const vslotShorthandName = NodeProcessor.getVslotShorthandName(child);
+      // vslot is guaranteed to exist since we have filtered
```
Comment (ang-zeyu) [inline packages/core/src/html/NodeProcessor.js:231]:
```suggestion
      
```
(all good if it dosen't exist too)

File: docs/userGuide/components/advanced.md [hunk 8]
```diff
@@ -36,6 +36,21 @@ You can define such a slot within the component by adding a `slot="attribute nam
 <variable name="highlightStyle">html</variable>
 </include>
 </div>
+
+<box type="tip">
+
+You may define a slot by using the shorthand notation `#attribute_name` as well!
```
Comment (ang-zeyu) [inline docs/userGuide/components/advanced.md:42]: let's encourage using this shiny new shorthand (list it upfront), then mention the deprecated syntax would still work too

**Review by ang-zeyu — COMMENTED** [review]
just a couple of nits:

File: packages/core/src/plugins/algolia.js [hunk 9]
```diff
@@ -24,9 +24,9 @@ function addNoIndexClasses(content) {
     'dropdown',
     'b-modal',
     'panel:not([expanded])',
-    'popover div[slot=content]',
-    'question div[slot=hint]',
-    'question div[slot=answer]',
+    'popover div["#content"]',
```
[the remark this answers is not in the document]
Comment (wxwxwxwx9) [reply packages/core/src/plugins/algolia.js]: Thanks for the spot! Have rectified this :-)

File: docs/userGuide/components/advanced.md (hunk 8 above)
Comment (wxwxwxwx9) [reply docs/userGuide/components/advanced.md]: Have made the changes! :-)

File: docs/userGuide/components/advanced.md [hunk 10]
```diff
@@ -36,6 +36,21 @@ You can define such a slot within the component by adding a `slot="attribute nam
 <variable name="highlightStyle">html</variable>
 </include>
 </div>
+
+<box type="info">
+
+You may define a slot by using `slot="attribute_name"` as well.
```
Comment (ang-zeyu) [inline docs/userGuide/components/advanced.md:42]:
```suggestion
You may define a slot by using `slot="slot_name"` as well. 
```

File: docs/userGuide/components/advanced.md [hunk 11]
```diff
@@ -11,14 +11,14 @@
 Most component attributes allow a richer form of formatting using slots, denoted by an attribute<strong>^\[S\]^</strong> superscript in the respective components' tables.
 In other cases, when the option is of type "Slot", only the slot option is available.
 
-You can define such a slot within the component by adding a `slot="attribute name"` attribute to any element within the slot.
+You can define such a slot within the component by adding a `#attribute_name` attribute to any element within the slot.
```
Comment (ang-zeyu) [inline docs/userGuide/components/advanced.md:14]:
```suggestion
You can define such a slot within the component by adding a `#slot_name` attribute to any element within the slot.
```
let's update the old one to make a little more sense

**Review by ang-zeyu — COMMENTED** [review]
Lgtm 👍

Let's update the tests too

**Comment by ang-zeyu** [comment]
> Update deprecated Vue slot syntax

Let's add a body too, describing the two types of changes here: external - shorthand for user, internal - .vue files

File: packages/cli/test/functional/test_site_algolia_plugin/expected/index.html [hunk 12]
```diff
@@ -81,16 +81,16 @@
...
     </popover>
     <popover effect="fade" header="Title" content="Content as attribute does not require `algolia-no-index` class" placement="top">
       <button class="btn btn-secondary">Trigger should not have `algolia-no-index` class</button></popover>
     <p><strong>Question hint and answer should have algolia-no-index class</strong></p>
     <question>
       Question should not have `algolia-no-index` class
-      <div slot="hint" class="algolia-no-index">Hint should have `algolia-no-index` class</div>
-      <div slot="answer" class="algolia-no-index">Answer should have `algolia-no-index` class</div>
+      <div slot="hint">Hint should have `algolia-no-index` class</div>
+      <div slot="answer">Answer should have `algolia-no-index` class</div>
```
Comment (wxwxwxwx9) [inline packages/cli/test/functional/test_site_algolia_plugin/expected/index.html:93]: Hmm, zeyu, I think this shouldn't be happening. The deprecated slot syntax should be transformed into the shorthand syntax and the node should be shifted one layer deeper. 

I'm currently looking into this to find out why it happens. I discovered that if I remove the <dropdown> from above, the deprecated slot syntax would be transformed properly. 

But I'm still investigating this :O

Comment (wxwxwxwx9) [reply packages/cli/test/functional/test_site_algolia_plugin/expected/index.html:93]: For this problem, I realised that it's `insertTemporaryStyles` that is causing the weird behaviour (any element that has slot after dropdown is not transformed properly). We can see many nodes of `<div class="temp-dropdown temp-dropdown-placeholder"></div>` being created before the `<dropdown>`. Can I check why is `insertTemporaryStyles` needed? 

Also, I just realised another problem for adding the `algolia-no-index` class. Currently, what we have for the selectors are 
`
    'popover div[\\#content]',
    'question div[\\#hint]',
    'question div[\\#answer]',
`

This would work for the following code:
`
<popover>
   <div #content>test</div>
</popover>
`

However, our processing actually shifts the node one layer deeper. So we would actually have the following instead:
`
<popover>
   <template #content>
      <div>test</div>
   </template>
</popover>
`

This means that our selectors would not work, and `algolia-no-index` class won't be added successfully. To overcome this, I think we may need to have the following instead:
`
    'popover template[\\#content]',
    'question template[\\#hint]',
    'question template[\\#answer]',
`

But I just wish to clarify if having `algolia-no-index` class on template html element would break anything? :O

Comment (wxwxwxwx9) [reply packages/cli/test/functional/test_site_algolia_plugin/expected/index.html:93]: > For this problem, I realised that it's `insertTemporaryStyles` that is causing the weird behaviour (any element that has slot after dropdown is not transformed properly). We can see many nodes of `<div class="temp-dropdown temp-dropdown-placeholder"></div>` being created before the `<dropdown>`.

I think we may be having a bug in the expected html file where we have many duplicated `<div class="temp-dropdown temp-dropdown-placeholder"></div>` generated above of `<dropdown>` to combat FOUC. 

From my debugging, I discovered that the problem stems from the function `insertTemporaryStyles` where we add the dropdown placeholders before the dropdown node itself:
```
    $dropdown.before(placeholder);
    $dropdown.prev().addClass(attributes.class).addClass(TEMP_DROPDOWN_PLACEHOLDER_CLASS);
```

I think this somehow messes up the processing of the nodes. All the nodes (e.g. popover, modal) that follows the dropdown would not get processed. I tested it out by checking if any nodes that are of the name "popover", "modal", "panel" are processed after the dropdown but it doesn't seem so. Instead, we are processing the "dropdown" node multiple times, in place of the rest of the nodes (e.g. "popover", "modal", "panel") for the test. 

I think it can be fixed by adding the placeholder after instead. Here's the modified code:
```
    $dropdown.after(placeholder);
    $dropdown.next().addClass(attributes.class).addClass(TEMP_DROPDOWN_PLACEHOLDER_CLASS);
```

I think the above should fix the problem and we can see that in the expected html file, the duplicated temporary dropdown placeholder are gone and we are left with only one of it (which should be the expected behaviour, correct me if I'm wrong).

Comment (ang-zeyu) [reply packages/cli/test/functional/test_site_algolia_plugin/expected/index.html:93]: > Otherwise, we can actually use a more specific css selector to place the algolia-no-index class on the div instead.

Not exactly sure how algolia does its indexing. To be safe, let's go with this instead since `template`s are erased by vue after compilation.

> I think we may be having a bug in the expected html file where we have many duplicated <div class="temp-dropdown temp-dropdown-placeholder"></div> generated above of <dropdown> to combat FOUC.

Thanks for investigating this! : )

> I think this somehow messes up the processing of the nodes. All the nodes (e.g. popover, modal) that follows the dropdown would not get processed.

To confirm -- `.after()` would result in a 'normal / fixed' processing order?

If so let's do that then, preferably in a separate PR

Comment (wxwxwxwx9) [reply packages/cli/test/functional/test_site_algolia_plugin/expected/index.html:93]: > To confirm -- `.after()` would result in a 'normal / fixed' processing order? (as a 'sibling' operation seems to be causing this)
> 
> If so let's do that then, preferably in a separate PR

Can I clarify about "normal / fixed" processing order? :O 

Suppose we have: 
```
<modal></modal>
<popover></popover>
<panel></panel>
```

You mean if it would be processed in the sequential order, right? Like `modal` -> `popover` -> `panel`?

If so, yup, .after() would give us the correct sequential processing order.

Comment (ang-zeyu) [reply packages/cli/test/functional/test_site_algolia_plugin/expected/index.html:93]: e.g.

```
// suppose these are the last 3 elements at the same level
<dropdown>
<div class="temp-dropdown-placeholder></div> // gets inserted here with after()
<panel> // this actually gets processed
```

Comment (wxwxwxwx9) [reply packages/cli/test/functional/test_site_algolia_plugin/expected/index.html:93]: Yup, that's the behaviour I'm seeing currently :-)

**Review by ang-zeyu — COMMENTED** [review]
a few more minor suggestions I missed earlier:

File: packages/core/src/html/NodeProcessor.js [hunk 13]
```diff
@@ -88,17 +89,35 @@ class NodeProcessor {
...
    * @param attribute Attribute name to process
    * @param isInline Whether to process the attribute with only inline markdown-it rules
    * @param slotName Name attribute of the <slot> element to insert, which defaults to the attribute name
    */
   _processAttributeWithoutOverride(node, attribute, isInline, slotName = attribute) {
     const hasAttributeSlot = node.children
-      && node.children.some(child => _.has(child.attribs, 'slot') && child.attribs.slot === slotName);
+      && node.children.some((child) => {
+        const vslotShorthandName = NodeProcessor.getVslotShorthandName(child);
+        return vslotShorthandName && vslotShorthandName === slotName;
```
[the remark this answers is not in the document]
Comment (wxwxwxwx9) [reply packages/core/src/html/NodeProcessor.js]: Yup, that's a good idea!

File: packages/core/src/html/NodeProcessor.js [hunk 14]
```diff
@@ -124,11 +143,11 @@ class NodeProcessor {
...
-        child.attribs['data-mb-slot-name'] = slot;
-        delete child.attribs.slot;
+      // Turns <div #content>... into <div data-mb-slot-name=content>...
+      const vslotShorthandName = NodeProcessor.getVslotShorthandName(child);
+      if (vslotShorthandName) {
+        child.attribs['data-mb-slot-name'] = vslotShorthandName;
+        delete child.attribs[`#${vslotShorthandName}`];
       }
       // similarly, need to transform templates to avoid Vue parsing
       if (child.name === 'template') {
```
[the remark this answers is not in the document]
Comment (wxwxwxwx9) [reply packages/core/src/html/NodeProcessor.js]: Ahh right, thanks for the catch!

**Review by ang-zeyu — APPROVED** [review]
Lgtm 👍 Thanks! @wxwxwxwx9
